File: fb/210.course-schedule-ii.cpp
// Time O(V+E)
// Space O(V+E)
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> neighbor;
        vector<int> degree(numCourses, 0);
        for(const auto& v : prerequisites) {
            degree[v[0]]++;
            neighbor[v[1]].push_back(v[0]);
        }
        
        queue<int> q;
        for(int i = 0; i < numCourses; i++) {
            if(degree[i] == 0)
                q.push(i);
        }
        
        vector<int> ans;
        while(!q.empty()) {
            int out = q.front();
            q.pop();
            ans.push_back(out);
            for(auto n : neighbor[out]) {
                degree[n]--;
                if(degree[n] == 0)
                    q.push(n);
            }
        }
        
        if(ans.size() == numCourses)
            return ans;
        return {};
    }
};

File: fb/210.course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_list = defaultdict(list)
        indegree = {}
        
        for dest, src in prerequisites:
            adj_list[src].append(dest)
            indegree[dest] = indegree.get(dest, 0) + 1
        
        topolocial_sorted_set = []
        zero_indegree_queue = deque([v for v in range(numCourses) if v not in indegree])
        
        while zero_indegree_queue:
            vertex = zero_indegree_queue.popleft()
            topolocial_sorted_set.append(vertex)
            if adj_list.get(vertex):
                for neighbor in adj_list[vertex]:
                    indegree[neighbor] -= 1
                    
                    if indegree[neighbor] == 0:
                        zero_indegree_queue.append(neighbor)
        
        return topolocial_sorted_set if len(topolocial_sorted_set) == numCourses else []
